### DISCUSSION TITLE
Correct way to recreate the ChatClient?

### DISCUSSION THREAD

**Discussion by jeff-huston** [GetStream/stream-chat-android#3971] [General]
We have two use cases we need to support:
1. Disconnect and then connect to a different Stream user ID while the app is running
2. When switching environments in our app, we need to disconnect and reconnect the ChatClient with a different Stream API key (since our backend environment needs to point to the same stream environment as our app does).

The first use case is the most important, but the second one also matters.

Currently, I'm using the following approach for both situations:
* Call `ChatClient.disconnect()`
* Rebuild a new `ChatClient` with `ChatClient.Builder` and connect using that one.

This seems to work functionally, but I do get this error in logcat:
`E/Chat: [ERROR] You have just re-initialized ChatClient, old configuration has been overridden [ERROR]`

Looking at `ChatClient`'s API surface, it seems like for Use Case 1 I could probably just call `disconnect()` and then `connectUser()` with the new user. But for Use Case 2 I think I do need to recreate the `ChatClient`. I'd like to follow the same codepath for both use cases, but can diverge if necessary.

Are there any drawbacks I should be aware of with my current approach, or a better alternative you would suggest?

**Comment by jeff-huston** [comment]
Note: I just noticed the build warning and changed to `chatClient.disconnect(flushPersistence = true).await()`, but this error log still appears.

**Comment by adasiewiczr** [comment]
Hey @hustoj2,
1. That is precisely the way to go. It would be best if you waited for `disconnect` completion and reconnect with a different `userId`. We are working on making this process more straightforward by introducing `switchUser` method which should be available in one of the upcoming releases.
2. This one is more tricky. Rebuilding `ChatClient` is a way to go but keep in mind the `ChatClient` instance is being used internally by other classes and components and all of them should be rebuilt. You should disconnect the user before recreating the new instance, as internally it will clear classes related to state, etc. Also, make sure to rebuild all UI-related things like VMs, etc. I'll add a task to investigate if we can improve this process and make sure it's described in our docs

**Reply by jeff-huston** [reply]
Thanks for the reply and guidance on how to safely recreate the `ChatClient` when needed. This helped.

**Reply by jeff-huston** [reply]
Note for future travelers: the `switchUser` method was in fact added in [v5.9.0](https://github.com/GetStream/stream-chat-android/releases/tag/v5.9.0) (though I haven't tried it yet in our app since our old way is still working)

**Reply by adasiewiczr** [reply]
we also added more documentation for [Handling User Connection](https://getstream.io/chat/docs/sdk/android/client/guides/handling-user-connection/)
